synapse/rest/client/sync.py

```diff
@@ -298,14 +298,6 @@ async def encode_response(
         if archived:
             response["rooms"][Membership.LEAVE] = archived
 
-        if sync_result.groups is not None:
-            if sync_result.groups.join:
-                response["groups"][Membership.JOIN] = sync_result.groups.join
-            if sync_result.groups.invite:
-                response["groups"][Membership.INVITE] = sync_result.groups.invite
-            if sync_result.groups.leave:
-                response["groups"][Membership.LEAVE] = sync_result.groups.leave
-
```

Comment:
I think this means `response["groups"]` will now always be an empty dict?

Reply:
It will not appear at all.

`response` is a `defaultdict(dict)`, so `response["groups"]` magically creates a `dict` if it didn't exist, which we then assigned other stuff into it.

Each removed line is the same as:

```python
response.setdefault("groups", {})[Membership.FOO] = sync_result.groups.foo
```

Personally I wish we used `defaultdict` in more places instead of that pattern. 😄 

(I don't see anywhere else we define `response["groups"]` -- did I miss one?)

Reply:
Ahh, that makes sense. I hadn't looked at the surrounding source code thoroughly (read: at all). Thanks for explaining.

I quite like the defaultdict pattern too. (As a curiosity, I note it's not infinitely recursive. That's probably a good thing, but I wonder if it's possible?)

Reply:
> (As a curiosity, I note it's not infinitely recursive. That's probably a good thing, but I wonder if it's possible?)

You can do something like the following:

```pycon
>>> def gen():
>>>     return defaultdict(gen)
>>> d = gen()
>>> d["foo"]["bar"]["baz"] = "woot"
>>> json.dumps(d)
'{"foo": {"bar": {"baz": "woot"}}}'
```